planning.py

```diff
@@ -12,3 +11,114 @@
 import time
 import bisect
 import string
+
+
+class Action(Expr):
+    """Subclasses Expr class to define an Action class that can refer
+    to actions of an existing Node or a HLA"""
+    def isNull(self):
+        """Returns True if the HLA is Empty"""
+        if self == None:
+            return True
+        else:
+            return False
+        
+    def Refinements(self,outcome,frontier):
+        """Refinement function on HLA using frontier and outcome
+        Steps = frontier    Precondition = outcome"""
+        X=Node(self).path()
+        R = []
+        for each in X:
+            if each.state == outcome:
+                R.append(each.action)        
+        for each in R:
+            R.append(prefix)
+            R.append(suffix)
+            for each in R:
+                n = Node.__init__(plan,state, plan.parent,plan.action)
+                frontier.append(n)
+              
+#_____________________________________________________________________________#
+
+    
+def hierarchicalSearch(problem):
+    """ Artificial Intelligence A Modern Approach Figure 11.5 function
+    HIERARCHICAL_SEARCH(problem,hierarchy) returns a solution or failure                  
+    frontier <- a FIFO queue with[Act] as the only element
+    loop do
+        if EMPTY?(frontier)then return failure
+        plan<-POP(frontier) //chooses the shallowest plan in frontier
+        hla<-the first HLA in plan,or null if none
+        prefix,suffix <-action before and after hla in plan
+        outcome <- RESULT(problem.INITIAL-STATE,prefix)
+        if hla is null then //so plan is primitive
+            if outcome statisfies.problem.GOAL then return plan
+        else for each sequence in REFINEMENTS(hla,outcome,hierarchy)do
+            frontier <-INSERT(APPEND(prefix,sequence,suffix),frontier)
+    Figure 11.5 A breadth-first implementation of hierarchical forward
+    planning search.The initial plan supplied to the algorithm is [Act]. 
+    The REFINEMENTS function returns a set of action sequences, one for 
+    each refinement of the HLA whose preconditions are satisfied by the
+    specified state, outcome.
+    author :Sarthak Mahapatra"""
+
+    current = Node(problem.initial)
+    frontier = FIFOQueue()
+    frontier.append(current)
+    while True:
+        if not frontier:
+            return failure()
```

Comment:
Easier to say `return None` here. No need to define `failure` as a function. And not right to return an empty list as failure, because an empty list is a valid plan.